File: src/test/java/quebec/virtualite/backend/services/utils/TestUtils.java
package quebec.virtualite.backend.services.utils;

import org.springframework.http.HttpStatus;
import org.springframework.validation.BindException;
import org.springframework.validation.beanvalidation.SpringValidatorAdapter;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.server.ResponseStatusException;

import javax.validation.ConstraintViolation;
import javax.validation.ValidatorFactory;
import javax.validation.executable.ExecutableValidator;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.List;
import java.util.Set;

import static javax.validation.Validation.buildDefaultValidatorFactory;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.validation.ValidationUtils.invokeValidator;
import static quebec.virtualite.utils.CollectionUtils.list;

public class TestUtils
{
    private TestUtils()
    {
    }

    public static <T> void assertInvalid(T dto)
    {
        assertThat(validateDTO(dto)).isGreaterThanOrEqualTo(1);
    }

    public static <T> void assertInvalid(T tested, String methodName, Object param)
    {
        assertInvalid(tested, methodName, list(param));
    }

    public static <T> void assertInvalid(T tested, String methodName, Object param1,
        Object param2)
    {
        assertInvalid(tested, methodName, list(param1, param2));
    }

    public static <T> void assertInvalid(T tested, String methodName, Object param1,
        Object param2, Object param3)
    {
        assertInvalid(tested, methodName, list(param1, param2, param3));
    }

    public static void assertStatus(Throwable exception, HttpStatus expectedStatus)
    {
        assertThat(exception)
            .isInstanceOf(ResponseStatusException.class)
            .hasFieldOrPropertyWithValue("status", expectedStatus);
    }

    public static <T> void assertValid(T dto)
    {
        assertThat(validateDTO(dto)).isEqualTo(0);
    }

    public static <T> void assertValid(T tested, String methodName, Object param)
    {
        assertValid(tested, methodName, list(param));
    }

    public static <T> void assertValid(T tested, String methodName, Object param1,
        Object param2)
    {
        assertValid(tested, methodName, list(param1, param2));
    }

    public static <T> void assertValid(T tested, String methodName, Object param1,
        Object param2, Object param3)
    {
        assertValid(tested, methodName, list(param1, param2, param3));
    }

    public static <T> String validateController(
        T tested,
        String methodName,
        List<Object> parameterValues)
    {
        try (ValidatorFactory factory = buildDefaultValidatorFactory())
        {
            ExecutableValidator executableValidator = factory
                .getValidator()
                .forExecutables();

            Method method = findMethod(tested, methodName);
            for (int i = 0; i < method.getParameterCount(); i++)
            {
                if (isRequestBody(method, i)
                    && parameterValues.get(i) == null)
                {
                    return "Param " + i + " - Can't have a null for a required RequestBody";
                }
            }

            try
            {
                Set<ConstraintViolation<T>> errors =
                    executableValidator
                        .validateParameters(tested, method, parameterValues.toArray());

                return errors.isEmpty()
                       ? ""
                       : "Validation errors: " + errors;
            }
            catch (Exception e)
            {
                return "Invalid argument type: " + e.getMessage();
            }
        }
    }

    public static <T> int validateDTO(T dto)
    {
        try (ValidatorFactory validatorFactory = buildDefaultValidatorFactory())
        {
            BindException errors = new BindException(dto, dto.getClass().getSimpleName());
            SpringValidatorAdapter validator = new SpringValidatorAdapter(validatorFactory.getValidator());

            invokeValidator(validator, dto, errors);
            return errors.getErrorCount();
        }
    }

    private static <T> void assertInvalid(T tested, String methodName, List<Object> params)
    {
        String message = validateController(tested, methodName, params);
        assertThat(message)
            .withFailMessage("Expecting an error (but didn't get any)")
            .isNotBlank();
    }

    private static <T> void assertValid(T tested, String methodName, List<Object> params)
    {
        String message = validateController(tested, methodName, params);
        assertThat(message)
            .withFailMessage(message)
            .isBlank();
    }

    private static <T> Method findMethod(T tested, String methodName)
    {
        for (Method method : tested.getClass().getMethods())
        {
            if (method.getName().equals(methodName))
                return method;
        }

        throw new AssertionError("Method " + methodName + " not found in class " + tested
            .getClass().getSimpleName());
    }

    private static boolean isRequestBody(Method method, int noParameter)
    {
        Parameter[] methodParameters = method.getParameters();
        RequestBody requestBody = methodParameters[noParameter]
            .getAnnotation(RequestBody.class);

        return requestBody != null && requestBody.required();
    }
}

File: src/test/java/quebec/virtualite/backend/services/utils/TestUtils.kt
package quebec.virtualite.backend.services.utils

import org.assertj.core.api.Assertions.assertThat
import org.springframework.http.HttpStatus
import org.springframework.validation.BindException
import org.springframework.validation.ValidationUtils.invokeValidator
import org.springframework.validation.beanvalidation.SpringValidatorAdapter
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.server.ResponseStatusException
import java.lang.reflect.Method
import javax.validation.ConstraintViolation
import javax.validation.Validation.buildDefaultValidatorFactory

object TestUtils
{
    fun assertInvalid(
        controller: Any,
        methodName: String,
        param: Any?
    )
    {
        assertInvalid(controller, methodName, listOf(param))
    }

    fun assertInvalid(
        controller: Any,
        methodName: String,
        param1: Any?,
        param2: Any?
    )
    {
        assertInvalid(controller, methodName, listOf(param1, param2))
    }

    fun assertInvalid(
        controller: Any,
        methodName: String,
        param1: Any?,
        param2: Any?,
        param3: Any?
    )
    {
        assertInvalid(controller, methodName, listOf(param1, param2, param3))
    }

    fun assertInvalid(dto: Any)
    {
        assertThat(validateDTO(dto))
            .withFailMessage("Validation should have failed (but didn't)")
            .isGreaterThanOrEqualTo(1)
    }

    fun assertStatus(exception: Throwable, expectedStatus: HttpStatus)
    {
        assertThat(exception)
            .isInstanceOf(ResponseStatusException::class.java)
            .hasFieldOrPropertyWithValue("status", expectedStatus)
    }

    fun assertValid(
        controller: Any,
        methodName: String,
        param: Any?
    )
    {
        assertValid(controller, methodName, listOf(param))
    }

    fun assertValid(
        controller: Any,
        methodName: String,
        param1: Any?,
        param2: Any?
    )
    {
        assertValid(controller, methodName, listOf(param1, param2))
    }

    fun assertValid(
        controller: Any,
        methodName: String,
        param1: Any?,
        param2: Any?,
        param3: Any?
    )
    {
        assertValid(controller, methodName, listOf(param1, param2, param3))
    }

    fun assertValid(dto: Any)
    {
        assertThat(validateDTO(dto))
            .withFailMessage("Validation failed")
            .isEqualTo(0)
    }

    fun validateController(
        controller: Any,
        methodName: String,
        parameterValues: List<Any?>
    ): String
    {
        buildDefaultValidatorFactory().use { factory ->

            val method: Method = findMethod(controller, methodName)

            if (searchForNullRequiredRequestBody(method, parameterValues) != null)
                return "Can't have a null for a required RequestBody"

            try
            {
                val errors: Set<ConstraintViolation<Any>> = factory
                    .validator
                    .forExecutables()
                    .validateParameters(controller, method, parameterValues.toTypedArray())

                return if (errors.isEmpty()) "" else "Validation errors: $errors"

            } catch (e: Exception)
            {
                return "Invalid argument type: " + e.message
            }
        }
    }

    fun validateDTO(dto: Any): Int
    {
        buildDefaultValidatorFactory().use { validatorFactory ->

            val errors = BindException(dto, dto.javaClass.simpleName)
            val validator = SpringValidatorAdapter(validatorFactory.validator)

            invokeValidator(validator, dto, errors)
            return errors.errorCount
        }
    }

    private fun assertInvalid(
        controller: Any,
        methodName: String,
        params: List<Any?>
    )
    {
        assertThat(validateController(controller, methodName, params))
            .withFailMessage("Expecting an error (but didn't get any)")
            .isNotBlank()
    }

    private fun assertValid(
        controller: Any,
        methodName: String,
        params: List<Any?>
    )
    {
        val message: String = validateController(controller, methodName, params)
        assertThat(message)
            .withFailMessage(message)
            .isBlank()
    }

    private fun findMethod(type: Any, methodName: String): Method
    {
        for (method in type.javaClass.getMethods())
        {
            if (method.name.equals(methodName))
                return method
        }

        throw AssertionError("Method $methodName not found in class ${type.javaClass.simpleName}")
    }

    private fun isRequiredRequestBody(method: Method, noParameter: Int): Boolean
    {
        val methodParameters = method.parameters
        val requestBody: RequestBody? = methodParameters[noParameter]
            .getAnnotation(RequestBody::class.java)

        return requestBody != null && requestBody.required
    }

    private fun searchForNullRequiredRequestBody(
        method: Method,
        parameterValues: List<Any?>
    ): Int?
    {
        for (i in 0..method.parameterCount - 1)
        {
            if (isRequiredRequestBody(method, i)
                && parameterValues[i] == null
            )
            {
                return i
            }
        }

        return null
    }
}
